APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+
+New contributors
+~~~~~~~~~~~~~~~~
+New contributors may be more discouraged by more steps. However:
+
+This *replaces* the existing PEP8 check, so that it is not "another step" To
+make it better, leave very clear explicit messages in the status checks on GH
+(not putting up a barrier) Scikit-learn had a few issues with contributors but
+not at a level that caused regret about the decision to adopt black formatting.
+One developer described the change as a "net benefit".
+
+Removing the humanity
+~~~~~~~~~~~~~~~~~~~~~
+Some developers may feel that individual style and humanity gets lost in
+autoformatting. We hope to convey that humanity and individuality is expressed
+more fully in the elegance and clarity of the code logic and implementation. In
+most cases the placement of white space and line breaks can be viewed as a
+mechanistic process that humans need not control.
+
+Exception Policy
+^^^^^^^^^^^^^^^^
+As described above, there are cases Black formatting makes the code
+substantially more difficult to read or understand.. These should be treated
+explicitly as exceptions, using the ``#fmt: off``/``#fmt: on`` escape hatch for
+multi-line exceptions, or ``#fmt: skip`` for single-line exceptions. **This is
+allowed if the code author and PR reviewer(s) agree that the code line or block
+should be excluded from Black formatting.**
+
+
+Branches and pull requests
+--------------------------
+
+- `Add configuration for black autoformatter <https://github.com/astropy/astropy/pull/13253>`_
+- `Apply black to modeling <https://github.com/astropy/astropy/pull/13254>`_
+
+Implementation
+--------------
+
+All Python code in ``astropy`` is formatted with Black_, using its default
+settings, that is, 88 characters per line and double quotes. Note that, by
+design, most of Black_ is inflexible, and there is not a way to pick-and-choose
+some elements and not others of Black's format.
+
+Implementing this change requires:
+
+1. Updating the `coding style`_:
+
+   * Adding documentation about Black, with a particular emphasis on the rationale expressed here and the exception policy described above.
+   * Adding documentation about using pre-commit, ala `pandas pre-commit`_.
+   * Document the black exception guidelines (as described in this APE)
+   * Removing other references to code formatting in the documentation.
+
+2. Updating flake8 configuration to be compatible with Black.
+
+   This is straightforward and well documented. It may even be removed if it
+   does not provide any added checks that ``black –check`` does not.
+
+3. Checking Black in CI.
+
+   * Add a CI check that runs ``black --check`` on the whole code base, failing
+     if it meets a violation
+   * Allow for the Black CI bot to fix formatting errors via documented single
+     line comments. This is an opt-in step which could be used by maintainers to
+     fix errors prior to merging. Having such an option available could lower
+     the barrier for new contributors.
```

Comment:
The usage of the [pre-commit.ci](https://pre-commit.ci/) bot proposed in astropy/astropy#13320, see the implementation plan https://github.com/astropy/astropy/issues/13320#issuecomment-1171488051 will perform all of these tasks for us. The function of this bot will be documented by astropy/astropy#13553 when we enable the bot.

Note the usage of pre-commit and the related pre-commit.ci bot are separate issues than adopting black.

Reply:
The pre-commit.ci bot was turned on today and documented with the merger of astropy/astropy#13353, so in the event that this APE is accepted, simply adding Black (or some other formatting tools) to the pre-commit configuration will be sufficient to enable opt-in automatic fixing.